tachyon/zk/plonk/vanishing/prover_vanishing_argument.h

```diff
@@ -0,0 +1,322 @@
+    // Obtain final h(X) polynomial
+    ExtendedPoly h_poly = ExtendedToCoeff(h_evals, prover->extended_domain());
+
+    // Truncate it to match the size of the quotient polynomial; the
+    // evaluation domain might be slightly larger than necessary because
+    // it always lies on a power-of-two boundary.
+    std::vector<F> h_coeffs = h_poly.coefficients().coefficients();
+    h_coeffs.resize(prover->domain()->size() * quotient_poly_degree, F::Zero());
+
+    // Compute commitments to each h(X) piece
+    const size_t kCommitmentNum = h_coeffs.size() / prover->pcs().N();
+
+    std::vector<bool> results = base::ParallelizeMapByChunkSize(
+        h_coeffs, prover->pcs().N(),
+        [prover](absl::Span<const F> h_piece, size_t chunk_index) {
+          return prover->Commit(h_piece);
+        });
+    if (std::any_of(results.begin(), results.end(),
+                    [](bool result) { return result == false; })) {
+      return false;
+    }
+
+    // FIXME(TomTaehoonKim): Remove this if possible.
+    std::vector<F> h_blinds;
+    h_blinds.reserve(kCommitmentNum);
+    for (size_t i = 0; i < kCommitmentNum; ++i) {
+      h_blinds.push_back(prover->blinder().Generate());
+    }
+
+    *constructed_out = Constructed<PCSTy>{
+        std::move(h_poly), std::move(h_blinds), std::move(*this)};
+    return true;
+  }
+
+ private:
+  Poly random_poly_;
+  F random_blind_;
+};
+
+template <typename PCSTy>
+class Constructed {
+ public:
+  using F = typename PCSTy::Field;
+  using Commitment = typename PCSTy::Commitment;
+  using ExtendedPoly = typename PCSTy::ExtendedPoly;
+
+  constexpr Constructed() = default;
+  constexpr Constructed(ExtendedPoly&& h_poly, std::vector<F>&& h_blinds,
+                        Committed<PCSTy>&& committed)
+      : h_poly_(std::move(h_poly)),
+        h_blinds_(std::move(h_blinds)),
+        committed_(std::move(committed)) {}
+
+  constexpr ExtendedPoly& h_poly() const { return h_poly; }
+  constexpr std::vector<F>& h_blinds() const { return h_blinds_; }
+  constexpr Committed<PCSTy>& committed() const { return committed_; }
+
+  bool Evaluate(const PCSTy& pcs, crypto::Challenge255<F>& x, const F& x_n,
+                crypto::TranscriptWriter<Commitment>* writer,
+                Evaluated<PCSTy>* evaluated_out) {
+    ExtendedPoly h_poly = ExtendedPoly::Zero();
+    auto h_chunks =
+        base::Chunked(h_poly_.coefficients().coefficients(), pcs.N());
+    auto h_pieces =
+        base::Map(h_chunks.begin(), h_chunks.end(),
+                  [](const absl::Span<const F>& h_piece) { return h_piece; });
+    for (absl::Span<const F> h_piece : base::Reversed(h_pieces)) {
+      std::vector<F> h_vec(h_piece.begin(), h_piece.end());
+      h_poly = h_poly * x_n +
+               ExtendedPoly(typename ExtendedPoly::Coefficients(h_vec));
+    }
+
+    F h_blind = std::accumulate(
+        h_blinds_.rbegin(), h_blinds_.rend(), F::Zero(),
+        [&x_n](const F& acc, const F& eval) { return acc * x_n + eval; });
```

Comment:
Can't we make it parallelized?

Reply:
I don't think we can since there is a dependency for the previous value when calculating.